pkg/skaffold/build/buildpacks/lifecycle.go

```diff
@@ -166,3 +164,49 @@ func envMap(env []string) map[string]string {
 
 	return kv
 }
+
+// fromRequiredArtifacts replaces the provided builder and run images with built images from the required artifacts if specified.
+func fromRequiredArtifacts(artifact *latest.BuildpackArtifact, r ArtifactResolver, deps []*latest.ArtifactDependency, pushImages bool) (string, string, bool) {
+	builderImage, runImage := artifact.Builder, artifact.RunImage
+	builderImageLocal, runImageLocal, noPull := false, false, false
+	var found bool
+	for _, d := range deps {
+		if builderImage == d.Alias {
+			builderImage, found = r.GetImageTag(d.ImageName)
+			if !found {
+				logrus.Fatalf("failed to resolve build result for required artifact %q", d.ImageName)
+			}
+			builderImageLocal = true
+		}
+		if runImage == d.Alias {
+			runImage, found = r.GetImageTag(d.ImageName)
+			if !found {
+				logrus.Fatalf("failed to resolve build result for required artifact %q", d.ImageName)
+			}
+			runImageLocal = true
+		}
```

Comment:
can these two code paths be combined into one? And then do the check at the end to either set `builderImageLocal = true` or `runImageLocal = true`

Reply:
not quite sure how though, can you suggest the code change you're thinking?

Reply:
On second look, maybe they can't, my bad 😅

Do you think the log messages should be differentiated to say either builder image or run image? Or do you think the user will have enough info from the ImageName?

Reply:
this is actually an impossible scenario. It'll only happen if we ever introduce a bug into the code.